src/wp-includes/class-wp-block-type.php

```diff
@@ -113,7 +113,8 @@ class WP_Block_Type {
 	 * Block variations.
 	 *
 	 * @since 5.8.0
-	 * @var array[]
+	 * @since 6.5.0 Added callback support.
+	 * @var callable|array[]
```

Comment:
This really feels like a breaking change. This is a public property on a class. I don't think we should do this.

Reply:
We could consider making this property private and use a `__get()` method to check and see if the block type was registered with a callback stored in a new `$variations_callback` property. If a callback was registered, the magic getter would return the result of the callable. If no callable is registered then it would return the property itself. We would need a `__set()` method to maintain the previous behavior of directly setting the variations property. However, this would mean that if a callback were available, it will be prioritized over any directly set value unless we somehow merged the callback value with any currently set values of the variations prop itself, which seems like it would get. messy.

Reply:
I agree with @spacedmonkey; this feels like a breaking change. Calling `$block_type_instance->variations` should return an array.